### PR TITLE
Added {logical_not, trace} refs, moved logical ops to use method overloads

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79002
* #79001
* __->__ #79000

### CODE REVIEW

**Comment by Chillee** [comment]
One thing I did in this PR that's questionable is porting `trace` from a decomp to a ref. Currently we don't have the `diag` ref implemented yet (so I need to xfail most of the ref tests), but I feel like it might be worth it in order to unify?

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -965,12 +967,25 @@ def _logical_and(a: TensorLikeType, b: TensorLikeType):
 )
 
 
+def _logical_not(a: TensorLikeType):
+    if not utils.is_boolean_dtype(a.dtype):
+        a = a != 0
```
Comment (mruberry) [inline torch/_refs/__init__.py:972]: return `a == 0`?

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -2419,6 +2434,10 @@ def equal(a: TensorLikeType, b: TensorLikeType) -> bool:
     return item(all(eq(a, b)))  # type: ignore[return-value]
 
 
-# populate the decomp table
+@register_decomposition(torch.ops.aten.trace)
+def trace(self: TensorLikeType) -> TensorLikeType:
+    return torch.sum(torch.diag(self))
```
Comment (mruberry) [inline torch/_refs/__init__.py:2439]: This is incorrect because diag accepts vectors and matrices but trace only accepts matrices

**Comment by mruberry** [comment]
Hey @Chillee! Looks good overall -- a few small comments for your review inline

**Comment by mruberry** [comment]
> One thing I did in this PR that's questionable is porting `trace` from a decomp to a ref. Currently we don't have the `diag` ref implemented yet (so I need to xfail most of the ref tests), but I feel like it might be worth it in order to unify?

Yeah don't worry about it

File: torch/_refs/__init__.py (hunk 2 above)
Comment (Chillee) [reply torch/_refs/__init__.py]: Added an error input for it (although it's unfortunately not tested now for ref implementations since diag isn't there).

**Review by ezyang — APPROVED** [review]

**Comment by mruberry** [comment]
Unfortunately test failures look related

**Comment by mruberry** [comment]
fyi @zasdfgbnm

**Pull request closed without merging**